### ISSUE TITLE
allow_clipping_in_normalization might be lossy in audio preprocessing

### ISSUE THREAD

**Issue by begeekmyfriend** [Rayhane-mamah/Tacotron-2#213]
I have printed the value of mel spectrogram before [normalization](https://github.com/Rayhane-mamah/Tacotron-2/blob/master/datasets/audio.py#L195) and I have found that the flooring and ceiling of those values exceed tha max absolute value set in `hparam.py`. I think it is lossy when we would like to invert mel spectrograms since many of the values have been changed manually and not true. For instance, we might get many negative and positive max absolute value (aka -4 or 4) as mel points which in fact do not exist in our dataset.
[mel_norm.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/2404898/mel_norm.zip)

**Comment by begeekmyfriend** [comment]
Below is the tarball including two evaluation wav files. One is spectrograms normalized ranged [-4, 4] and the other ranged [0, 1]. It sounds clearly that the one ranged [-4, 4] is much better than the other ranged [0,1]. `allow_clipping_in_normalization` has been disabled in both of two testing in preprocessing.
[[-4,4]vs[0,1].zip](https://github.com/Rayhane-mamah/Tacotron-2/files/2417788/-4.4.vs.0.1.zip)
The reason might be that [-4, 4] includes nagative featuring values. We can print `np.min(D)` and `np.max(D)` in both `inv_linear_spectrogram` and `inv_mel_spectrogram` method and find that many of the predicted spectrograms values are nagative. So featuring values ranged [-4, 4] might have high degree of reduction. `symmetric_mels = True` is necessary in my humble opinion.

**Comment by Liujingxiu23** [comment]
@begeekmyfriend Do you have any test to verify "I think the it better normalize abs(stft)."?

**Comment by begeekmyfriend** [comment]
That is from my [fork](https://github.com/begeekmyfriend/Tacotron-2/blob/mandarin-griffin-lim/hparams.py#L45-L48)
[mandarin_griffin_lim.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3102538/mandarin_griffin_lim.zip)

**Comment by carlfm01** [comment]
I would really appreciate if you @begeekmyfriend can help me with a hint of the source of my issue, please refer to : https://github.com/MlWoo/LPCNet/issues/3#issuecomment-529259284
 I'm trying to use taco2 and LPCNet, I'm a bit lost here, as clipping and normalization happens at preprocessing, I have no idea yet where my LCPNet are being chopped, I'm feeding them just as @MlWoo examples. Am I missing something?

Thanks.

**Comment by begeekmyfriend** [comment]
I did not achieve good results on T2 + LPCNet (original version) either. It seems there are something to do in the source of LPCNet vocoder. However I have got better results on WaveRNN vocoder. [Here](https://www.ximalaya.com/keji/28399446/208462208) is the production.

**Comment by carlfm01** [comment]
with the original me neither, using mlwoo  dumb_data.c adaptation here:
https://github.com/MlWoo/LPCNet/commit/7f17e2bc39e19b74a090fb3fac312eef7d6635f7#diff-1a1712f7f46a2a1b3317ae3e66506cc2

looks promising, by changing max_abs_value to 30 at inference improved a lot, something is still changing the prediction. I see not all the forks contains the same mechanism of clipping, I'll continue the research from there.
Listen:
[audios.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3593276/audios.zip)

It still looks like chopped from 800Hz +

And yes, WaveRNN sounds really good, the issue is that I need at the end use low compute devices.

**Comment by carlfm01** [comment]
Hello @begeekmyfriend I've managed to train LPCNet+Tacotron2 successfully applying few changes to tacotron 2:
It aligns with just 5k steps:
![step-5000-align](https://user-images.githubusercontent.com/32177100/65214794-87b8cd00-da68-11e9-9711-66dc44d90028.png)

25k steps:
![step-25000-align](https://user-images.githubusercontent.com/32177100/65214885-d0708600-da68-11e9-84cf-47dc8368e6dc.png)


audios:
[audios.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3629488/audios.zip)


See https://github.com/carlfm01/Tacotron-2/commits/spanish for the changes.

**Comment by begeekmyfriend** [comment]
I have got my own [branch](https://github.com/begeekmyfriend/Tacotron-2/tree/mandarin-lpc) but the evaluations is not as good as that of WaveRNN. So I lay aside it.

**Comment by carlfm01** [comment]
I see that your branch predicts 55d, we only need to predict 20d, according to the author LPCNet can estimate the LPC features out of the prediction from tacotron2.

https://github.com/mozilla/LPCNet/issues/4#issuecomment-455193542

Unfortunately I've failed  training WaveRNN, so I'm afraid I can't say which is better. Me as a native Spanish speaker can say that the quality of the generated voice are close or almost the same as the source. The issue with this voice is that it is really flat, so I dont noticed expressive speech. Right now I'm cutting a more expresive voice and see if this aproach is capable of geenrating expresive speech.

**Comment by begeekmyfriend** [comment]
My WaveRNN fork has been provided [here](https://github.com/begeekmyfriend/WaveRNN) with preprocessing in my [T2 fork](https://github.com/begeekmyfriend/Tacotron-2/blob/mandarin-mel-prediction/datasets/preprocessor.py). Copy the `training_data/audio` as `quant` and GTA mel as `gta` in WaveRNN `data` directory and just go on.

**Comment by carlfm01** [comment]
Thanks, did not knew about your WaveRNN fork. How much time it took for you to train WaveRNN?

**Comment by begeekmyfriend** [comment]
It takes about 5 days for 1000K steps in all. But the evaluations can be good enough within 400K steps in my experience.

**Comment by carlfm01** [comment]
> It takes about 5 days for 1000K steps in all.

How much data and GPU used? Thanks for your answers.

**Comment by begeekmyfriend** [comment]
In T2 repo
```shell
python synthesize.py --mode 'synthesis'
```
and copy the `tacotron_output/gta` as `gta` and `training_data/audio` as `quant` in WaveRNN root path where there are same number of npy files in both of two directory.
Single GPU is enough for the default hyperparameters.

**Comment by carlfm01** [comment]
> and copy the `tacotron_output/gta` as `gta` and `training_data/audio` as `quant` in WaveRNN root path where there are same number of npy files in both of two directory.

I understand thanks, 

> Single GPU is enough for the default hyperparameters.

About GPU, I meant which GPU took 5 days, a V100, a P100, or 1080?

**Comment by begeekmyfriend** [comment]
1080ti or 2080ti with 11GB memory is enough for 5 days.

**Comment by carlfm01** [comment]
Ok, thanks, good info to use it as reference.

**Comment by lyz04551** [comment]
> 1080ti or 2080ti with 11GB memory is enough for 5 days.

Can your wavernn model be real-time? I trained a good wavernn model 1200sample / s, what is the improvement in speed?

**Comment by CharlesSL** [comment]
@begeekmyfriend Hi, your fork about tacotron and wavernn is very helpful to my TTS work. Forgive my asking, are you familiar to ASR?

**Comment by begeekmyfriend** [comment]
My colleague takes the responsibility of that with espnet. I know little about it.

**Comment by wizardk** [comment]
@begeekmyfriend Hi, what about the inference speed on CPU of WaveRNN?

**Comment by begeekmyfriend** [comment]
@wizardk You'd better rewrite WaveRNN in C to test it on CPU.

**Comment by wizardk** [comment]
@begeekmyfriend Thanks for your suggestion. I have tested waveglow on CPU and it is pretty slow. Did you test wavernn on CPU already? I just want to make a reference beforehand.

**Comment by begeekmyfriend** [comment]
Not yet. There is a lot of work to do with rewriting in C. And I am afraid it would not open for public since it is a business behavior.

**Comment by wizardk** [comment]
@begeekmyfriend Thanks, I just want to know whether the WaveRNN's real time rate is lower than 1.

**Comment by m-toman** [comment]
You can check out this one
https://github.com/geneing/WaveRNN-Pytorch

Faster than realtime only with significant sparsification, with a big hit on quality.
At least in my experiments.

**Comment by wizardk** [comment]
@m-toman Thanks, that helps me a lot.
